modules/tensor_mechanics/src/materials/abaqus/AbaqusUmatMaterial.C

```diff
@@ -0,0 +1,262 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AbaqusUmatMaterial.h"
+#include "Factory.h"
+#include "MooseMesh.h"
+#include "RankTwoTensor.h"
+#include "RankFourTensor.h"
+
+#ifdef LIBMESH_HAVE_DLOPEN
+#include <dlfcn.h>
+#endif
+
+#define QUOTE(macro) stringifyName(macro)
+
+registerMooseObject("TensorMechanicsApp", AbaqusUmatMaterial);
+
+InputParameters
+AbaqusUmatMaterial::validParams()
+{
+  InputParameters params = ComputeStressBase::validParams();
+  params.addClassDescription("Coupling material to use Abaqus UMAT models in MOOSE");
+  params.addRequiredParam<FileName>(
+      "plugin", "The path to the compiled dynamic library for the plugin you want to use");
+  params.addRequiredParam<std::vector<Real>>("mechanical_constants",
+                                             "Mechanical Material Properties");
+  params.addParam<std::vector<Real>>("thermal_constants", "Thermal Material Properties");
+  params.addRequiredParam<unsigned int>("num_state_vars",
+                                        "The number of state variables this UMAT is going to use");
+  return params;
+}
+
+AbaqusUmatMaterial::AbaqusUmatMaterial(const InputParameters & parameters)
+  : ComputeStressBase(parameters),
+    _plugin(getParam<FileName>("plugin")),
+    _mechanical_constants(getParam<std::vector<Real>>("mechanical_constants")),
+    _thermal_constants(getParam<std::vector<Real>>("thermal_constants")),
+    _num_state_vars(getParam<unsigned int>("num_state_vars")),
+    _aqCELENT(1.0), // TODO: figure out what this really should be
```

Comment:
Maybe the comment should be something like "TODO: Compute the actual element characteristic length and provide that instead of 1"

Also, we should make a note of this and any other limitations like this in the .md for this class.

Reply:
Can we use `_current_elem->hmax()` here?

Reply:
Cool, I didn't even realize that we had `hmax()`.  I'm not sure whether that's the same as the characteristic length or not. My guess is that they take the cube root of the volume or something, which would also be easy to provide. I imagine that the Abaqus documentation says somewhere what the characteristic length is.

Reply:
"Characteristic element length, which is a typical length of a line across an element for a first-order element; it is half of the same typical length for a second-order element. For beams and trusses it is a characteristic length along the element axis. For membranes and shells it is a characteristic length in the reference surface. For axisymmetric elements it is a characteristic length in the  plane only. For cohesive elements it is equal to the constitutive thickness."